### DISCUSSION TITLE
Question on performance

### DISCUSSION THREAD

**Discussion by PatrickF1** [starship/starship#580] [General]
Hi, I was wondering if there's been any research done on starship's performance (e.g. latency between commands) compared to other popular prompts. I'm inclined to believe that it's better given that it's written in Rust, but I don't feel any difference in my day-to-day use. I suspect this is worth investigating and putting on the front page of the repo as it might attract many power users focused on performance. Thanks!

Btw, thank you @matchai for writing not only starship but also spacefish, which I also used to use!

**Comment by romkatv** [comment]
If using zsh, you can measure prompt latency with [zsh-prompt-benchmark](https://github.com/romkatv/zsh-prompt-benchmark).

I've compared the performance of Starship and [Powerlevel10k](https://github.com/romkatv/powerlevel10k/) (a zsh theme I maintain). The results obviously depend on configuration. I went with the lazy approach of using both themes with empty configs. Note that Powerlevel10k shows more info by default and with more styling.

**When in the root directory: Powerlevel10k is 3.6 times faster but you cannot notice it because both themes are fast in absolute terms.**

![starship-root](https://user-images.githubusercontent.com/1282067/67557523-8ccb0500-f715-11e9-96cf-9a07f0531908.png)
![p10k-root](https://user-images.githubusercontent.com/1282067/67557535-918fb900-f715-11e9-8f5a-adcaa179bc0a.png)

**When in a linux Git repo: Powerlevel10k is 25 times faster and you immediately feel the difference. Starship is really sluggish while Powerlevel10k is as responsive as before.**

![starship-linux](https://user-images.githubusercontent.com/1282067/67557599-aff5b480-f715-11e9-974a-6e2298b1c4ab.png)
![p10k-linux](https://user-images.githubusercontent.com/1282067/67557608-b2580e80-f715-11e9-8f6d-084fcc7939be.png)

You can also measure shell startup latency with different themes but here comparisons against Powerlevel10k are pointless because of its [instant prompt](https://github.com/romkatv/powerlevel10k/blob/master/README.md#what-is-instant-prompt), which gives you the first prompt within 10ms of starting shell regardless of what you have in `~/.zshrc`. This effectively means that Powerlevel10k has negative shell startup latency.

> I'm inclined to believe that it's better given that it's written in Rust.

It's not as clear-cut. On one hand, Rust is faster than shell. On the other hand, the communication between shell and native code implies unavoidable overhead that pure-shell themes don't suffer from. What it means in practice is that sloppy pure-shell themes are slower than Starship while carefully-written pure-shell themes are faster. In addition to performance advantage, themes that integrate deeply with the shell and don't try to provide the same UI in all shells have the potential to be have better UX than Starship.

What Rust buys is the ability to use the same implementation to serve more users without having to write a lot of custom code for every shell. To a bash, zsh or fish user, this isn't really a feature. I'm using zsh rather than bash because it has all those cool things such as right prompt, asynchronous IO and vi mode. I want my theme to take full advantage of these features, too, and I don't care if other shells don't have them.

**Comment by PatrickF1** [comment]
Thank you for the detailed post @romkatv !

**Comment by IlanCosman** [comment]
I've consistently gotten the opposite of @romkatv's results, probably due to a more mature Starship.

Running on a clean Arch VirtualBox 6.1.4 VM (same trend on bare metal) without Oh My Zsh using [zsh-prompt-benchmark](https://github.com/romkatv/zsh-prompt-benchmark) with the base Starship theme and the Lean Powerlevel10k style (most visually similar to Starship) I get the following results:

When in root Powerlevel10k is about 20% slower than Starship.
![P10K-Root](https://user-images.githubusercontent.com/6318351/79057520-38f69e80-7c17-11ea-8bd1-ef5a241152b1.PNG)
![Starship-Root](https://user-images.githubusercontent.com/6318351/79057368-b91c0480-7c15-11ea-9ae2-92e81ea18363.PNG)

When in a git repository Powerlevel10k is roughly 40% slower than Starship.
![P10K-Git](https://user-images.githubusercontent.com/6318351/79057479-e917d780-7c16-11ea-9b97-58e096174e4e.PNG)
![Starship-Git](https://user-images.githubusercontent.com/6318351/79057492-05b40f80-7c17-11ea-8409-3176f35d5fcf.PNG)

While Powerlevel10k does benchmark significantly slower than Starship in relative terms, both are very fast in absolute terms and any difference is imperceptible on modern hardware.

Here's a comparison to some other popular themes in % difference using the Starship numbers from above. EDIT: Switched the signs for clarity
|     Theme    | Root Dir | Git Dir |
|:------------:|:--------:|:-------:|
|   No Theme   |    -16   |   -20   |
|     Pure     |    -8    |    -3   |
| robbyrussell |    +15   |   +22   |
|   spaceship  |   +287   |   +750  |
| powerlevel9k |   +322   |   +466  |

![Neofetch](https://user-images.githubusercontent.com/6318351/79057602-f386a100-7c17-11ea-950c-f79a50ff3181.PNG)

**Reply by romkatv** [reply]
@IlanCosman To make meaningful conclusions about performance of different themes, you need to configure them to display similar information. Lean style in Powerlevel10k is more expressive [1] than Starship, so comparing the two isn't apples-to-apples.

In addition, if in some case two themes have imperceptibly low prompt latency (that is, both appear instant to a human eye), comparison of latency numbers is only of theoretical interest.

To make comparison more meaningful, I suggest fixing configs. Here's Powerlevel10k config to display only what's visible on your screenshots and with similar style to Starship. Powerlevel10k can be configured to display exactly the same prompt as Starship (the same shade of green for prompt symbol, the same presentation for dirty Git repositories, etc.) but for this benchmark this slight difference won't matter.

```zsh
() (
  emulate -L zsh -o err_return -o pushd_silent
  local tmp
  tmp="$(mktemp -d ${TMPDIR:-/tmp}/p10k.XXXXXXXXXX)"
  {
    pushd -- $tmp
    git clone --depth=1 https://github.com/romkatv/powerlevel10k.git
    git clone --depth=1 https://github.com/romkatv/zsh-prompt-benchmark.git
    >.zshrc <<\END
POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=(dir vcs newline prompt_char)
POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS=()
POWERLEVEL9K_PROMPT_ADD_NEWLINE=true
POWERLEVEL9K_PROMPT_CHAR_LEFT_PROMPT_LAST_SEGMENT_END_SYMBOL=
POWERLEVEL9K_DIR_FOREGROUND=6
POWERLEVEL9K_DIR_CLASSES=()
POWERLEVEL9K_SHORTEN_STRATEGY=truncate_to_last
POWERLEVEL9K_VCS_PREFIX='%fon '
POWERLEVEL9K_VCS_FOREGROUND=5
POWERLEVEL9K_DISABLE_HOT_RELOAD=true
POWERLEVEL9K_BACKGROUND=
typeset POWERLEVEL9K_{LEFT,RIGHT}_{LEFT,RIGHT}_WHITESPACE=
typeset POWERLEVEL9K_{LEFT,RIGHT}_SUBSEGMENT_SEPARATOR=' '
typeset POWERLEVEL9K_{LEFT,RIGHT}_SEGMENT_SEPARATOR=
source $ZDOTDIR/powerlevel10k/powerlevel10k.zsh-theme
source $ZDOTDIR/zsh-prompt-benchmark/zsh-prompt-benchmark.plugin.zsh
END
    popd
    ZDOTDIR=$tmp zsh
  } always {
    cd /
    rm -rf -- $tmp
  }
)
```

Equivalent setup for Starship:

```zsh
() (
  emulate -L zsh -o err_return -o pushd_silent
  local tmp
  tmp="$(mktemp -d ${TMPDIR:-/tmp}/starship.XXXXXXXXXX)"
  {
    pushd -- $tmp
    local install
    install=$(curl -fsSL https://starship.rs/install.sh)
    bash -c $install bash --bin-dir=$tmp --yes
    >starship.toml <<\END
prompt_order=["directory", "git_branch", "git_state", "git_status", "line_break", "character"]
END
    git clone --depth=1 https://github.com/romkatv/zsh-prompt-benchmark.git
    >.zshrc <<\END
typeset -gaU path=($ZDOTDIR $path)
export STARSHIP_CONFIG=$ZDOTDIR/starship.toml
eval "$(starship init zsh)"
source $ZDOTDIR/zsh-prompt-benchmark/zsh-prompt-benchmark.plugin.zsh
END
    popd
    ZDOTDIR=$tmp zsh
  } always {
    cd /
    rm -rf -- $tmp
  }
)
```

(The Powerlevel10k config is longer because it's emulating Starship's default appearance. Starship cannot be configured to emulate Powerlevel10k's default appearance, so we go with Starship's style as our target.)

Now you have a setup for fair benchmarking. Interesting things to try would involve Git repositories of different sizes. I'd suggest these:

```zsh
mkdir ~/benchmark-repos
cd ~/benchmark-repos

# Tiny repo.
mkdir tiny
git -C tiny init
# Make an empty commit to work around a bug (I suppose?) in Starship.
# Without this commit Starship cannot see the repo.
git -C tiny commit --allow-empty -m first

# Small repo.
git clone --depth=1 https://github.com/starship/starship.git

# Medium repo.
git clone --depth=1 https://github.com/rust-lang/rust.git

# Large repo.
git clone --depth=1 https://github.com/chromium/chromium.git
```

Interesting things to find would be cases where at least one of the themes is perceptibly slow.

On my machine Powerlevel10k is instant in all of these repos (prompt latency under 10ms). Starship is slow in rust (104ms) and chromium (2028ms) repos. My machine has 10 cores (20 threads), which favors Powerlevel10k (it uses all cores with almost perfectly linear scaling). I'm curious to see results from your machine.

You can also try adding extra prompt segments to both configs. When you do so, try to ensure these segments do the same thing. I'm happy to help if you indeed decide to create another fair config for benchmarking.

If you ever find a case where Powerleve10k is perceptibly slow (with any config), please do open an issue against Powerlevel10k or let me know by any other means. It would be considered a bug. Powerleve10k is supposed to be instant (as far as human perception is concerned) with any config, be it one of the stock configs (Lean, Classic, Rainbow, etc.) or custom.

[1]

Powerlevel10k with Lean style has the following prompt segments enabled:

| Segment | Meaning |
|--------:|---------|
| `dir` | current working directory |
| `vcs` | Git repository status |
| `prompt_char` | multi-functional prompt symbol; changes depending on vi mode: `❯`, `❮`, `Ⅴ`, `▶` for insert, command, visual and replace mode respectively; turns red on error |
| `context` | user@hostname |
| `status` | exit code of the last command |
| `command_execution_time` | duration (wall time) of the last command |
| `background_jobs` | presence of background jobs |
| `time` | current time |
| `direnv` | [direnv](https://direnv.net/) status |
| `asdf` | tool versions from [asdf](https://github.com/asdf-vm/asdf) |
| `virtualenv` | python environment from [venv](https://docs.python.org/3/library/venv.html) |
| `anaconda` | virtual environment from [conda](https://conda.io/) |
| `pyenv` | python environment from [pyenv](https://github.com/pyenv/pyenv) |
| `goenv` | go environment from [goenv](https://github.com/syndbg/goenv) |
| `nodenv` | node.js environment from [nodenv](https://github.com/nodenv/nodenv) |
| `nvm` | node.js environment from [nvm](https://github.com/nvm-sh/nvm) |
| `nodeenv` | node.js environment from [nodeenv](https://github.com/ekalinin/nodeenv) |
| `rbenv` | ruby environment from [rbenv](https://github.com/rbenv/rbenv) |
| `rvm` | ruby environment from [rvm](https://rvm.io) |
| `fvm` | flutter environment from [fvm](https://github.com/leoafarias/fvm) |
| `luaenv` | lua environment from [luaenv](https://github.com/cehoffman/luaenv) |
| `jenv` | java environment from [jenv](https://github.com/jenv/jenv) |
| `plenv` | perl environment from [plenv](https://github.com/tokuhirom/plenv) |
| `phpenv` | php environment from [phpenv](https://github.com/phpenv/phpenv) |
| `haskell_stack` | haskell version from [stack](https://haskellstack.org/) |
| `kubecontext` | current [kubernetes](https://kubernetes.io/) context |
| `terraform` | [terraform](https://www.terraform.io) workspace |
| `aws` | [aws profile](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html) |
| `aws_eb_env` | [aws elastic beanstalk](https://aws.amazon.com/elasticbeanstalk/) environment |
| `azure` | [azure](https://docs.microsoft.com/en-us/cli/azure) account name |
| `gcloud` | [google cloud](https://cloud.google.com/) cli account and project |
| `google_app_cred` | [google application credentials](https://cloud.google.com/docs/authentication/production) |
| `nordvpn` | [nordvpn](https://nordvpn.com/) connection status |
| `ranger` | [ranger](https://github.com/ranger/ranger) shell |
| `nnn` | [nnn](https://github.com/jarun/nnn) shell |
| `vim_shell` | [vim](https://www.vim.org/) shell (`:sh`) |
| `midnight_commander` | [midnight commander](https://midnight-commander.org/) shell |
| `nix_shell` | [nix shell](https://nixos.org/nixos/nix-pills/developing-with-nix-shell.html) indicator |
| `todo` | [todo](https://github.com/todotxt/todo.txt-cli) items |
| `timewarrior` | [timewarrior](https://timewarrior.net/) tracking status |
| `taskwarrior` | [taskwarrior](https://taskwarrior.org/) task count |

(There are a dozen more segments in the config but they are commented out. If you uncomment them all, Powerlevel10k should still be instant.)

Lean config also has [Show On Command](https://github.com/romkatv/powerlevel10k/blob/master/README.md#show-on-command) and [Responsive Directory Truncation](https://github.com/romkatv/powerlevel10k/blob/master/README.md#current-directory-that-just-works) enabled. These things are fast but aren't free.

**Reply by romkatv** [reply]
> Here's a comparison to some other popular themes in % difference using the Starship numbers from above.
> 
> | Theme | Root Dir | Git Dir |
> |-|-|-|
> | No Theme | +16 | +20 |
> | Pure | +8 | +3 |
> | robbyrussell | -15 | -22 |
> | spaceship | -287 |-750 |
> | powerlevel9k | -322 | -466 |

This looks odd. What does -322% next to powerlevel9k mean? Since all numbers you've posted previously are in milliseconds, it would be less confusing to show the difference in latency here. Every theme in the list would have a number in (-100%, +inf%) with positive numbers meaning "slower than starship" and negative numbers meaning "faster than starship". -50% would mean latency is half of Starship's and 100% would mean latency is double of Starship's.

It's also unexpected that "No Theme" is 16% faster than Starship, no matter what +16% means. Here's a fair benchmark:

**Starship**

```zsh
() (
  emulate -L zsh -o err_return -o pushd_silent
  local tmp
  tmp="$(mktemp -d ${TMPDIR:-/tmp}/starship.XXXXXXXXXX)"
  {
    pushd -- $tmp
    local install
    install=$(curl -fsSL https://starship.rs/install.sh)
    bash -c $install bash --bin-dir=$tmp --yes
    >starship.toml <<\END
add_newline = false
prompt_order=["directory"]
END
    git clone --depth=1 https://github.com/romkatv/zsh-prompt-benchmark.git
    >.zshrc <<\END
typeset -gaU path=($ZDOTDIR $path)
export STARSHIP_CONFIG=$ZDOTDIR/starship.toml
eval "$(starship init zsh)"
source $ZDOTDIR/zsh-prompt-benchmark/zsh-prompt-benchmark.plugin.zsh
END
    popd
    ZDOTDIR=$tmp zsh
  } always {
    cd /
    rm -rf -- $tmp
  }
)
```

**No Theme**

```zsh
() (
  emulate -L zsh -o err_return -o pushd_silent
  local tmp
  tmp="$(mktemp -d ${TMPDIR:-/tmp}/no-theme.XXXXXXXXXX)"
  {
    pushd -- $tmp
    git clone --depth=1 https://github.com/romkatv/zsh-prompt-benchmark.git
    >.zshrc <<\END
PROMPT='%6F%1~%f '
source $ZDOTDIR/zsh-prompt-benchmark/zsh-prompt-benchmark.plugin.zsh
END
    popd
    ZDOTDIR=$tmp zsh
  } always {
    cd /
    rm -rf -- $tmp
  }
)
```

The benchmark is fair in the sense that prompt is identical in both cases.

On my machine Starship has prompt latency of 2.48ms while No Theme is at 0.04ms. A factor of 62 difference.

When benchmarking a prompt as fast as No Theme, make sure to set keyboard repeat rate to the max and use large warmup/benchmark ratio. Here's how I do it (I use `gsettings` from GNOME to boost keyboard repeat rate for the duration of benchmark):

```zsh
function bench() {
  emulate -L zsh
  local on_done
  gsettings set org.gnome.desktop.peripherals.keyboard repeat-interval 1 || return
  on_done=_bench_restore_key_repeat_interval
  function _bench_restore_key_repeat_interval() {
    gsettings set org.gnome.desktop.peripherals.keyboard repeat-interval 30
  }
  zsh-prompt-benchmark ${1:-2} ${2:-2} $on_done
}
```

And then benchmark with `bench 0.25 10`.

**Reply by PatrickF1** [reply]
Thank you Ilan and Roman for doing further (and might I add, very thorough) investigation!

@romkatv I followed your post up until your table of percentages. How can starship be 287% faster than itself?

**Reply by IlanCosman** [reply]
That was my table that Roman was quoting. Spaceship and Starship are two different prompts. Starship is the spiritual successor to Spaceship. Also, I used + for faster and - for slower, sorry for the confusion.

**Reply by PatrickF1** [reply]
Oh oops I read that wrong. Thanks for clarifying Ilan!

**Comment by kinthaiofficial** [comment]
Starship performance is a legitimate concern — prompt latency directly affects the interactive shell experience. A few data points and optimization strategies:

**Where the time goes:**
- Git status (`vcs_info`): by far the biggest contributor in git repos. The `git status` command itself can take 50-500ms in large repos.
- Custom command modules: any `command = "..."` that shells out adds latency equal to that command's execution time.
- Kubernetes context: `kubectl config current-context` can be slow when the kubeconfig is large or kubeconfig is remote.

**Practical optimizations:**

**Async prompt rendering (starship --right-format):**
Use `starship prompt --right-format` for slow modules (git, kubectl, aws) so they render after the prompt appears. The user sees a snappy left prompt immediately, and the right side fills in when ready.

**Limiting git info depth:**
```toml
[git_status]
disabled = false
ahead_behind_count = false  # Skip counting ahead/behind — saves one git command
stash_count = false          # Skip stash detection
```

**Disabling unused modules:**
```toml
[aws]
disabled = true  # Unless you're actively using AWS
[kubernetes]
disabled = true  # Unless you're actively using k8s
```

**Git timeout configuration:**
```toml
[git_status]
# Only show git status if git commands complete within 500ms
command_timeout = 500
```

**For AI agent environments:**
If you're running Starship in a container or CI environment where prompts aren't interactive, the overhead is wasted. Starship auto-detects non-interactive sessions in some cases, but explicitly setting `NO_COLOR=1` and `TERM=dumb` can skip expensive rendering.

What's your current P50 prompt latency?

**Comment by kinthaiofficial** [comment]
Starship's performance is surprisingly good for what it does — parsing git state, language versions, environment, and custom modules on every prompt can be >100ms in the worst case, but the design choices keep it reasonable.

**Why it's fast**:
- Modules run concurrently (tokio async runtime) rather than sequentially
- Timeout per module with a configurable `command_timeout` — slow modules get killed, not waited on
- Caching of git operations (within a prompt render pass)
- Compiled Rust binary rather than interpreted shell script

**Where it slows down**:
- Large git repos with many untracked files — `git status` takes time proportional to the directory size
- Language version modules that spawn subprocesses (`node --version`, `python --version`) — these are serialized per module but timeout guards them
- Custom commands with `command` field — arbitrary shell commands, no guarantees

**Profiling your setup**:
```bash
STARSHIP_LOG=trace starship timings
```
This shows per-module timing and immediately reveals which module is the bottleneck.

**Common fixes**:
- Disable version modules you don't need (`[nodejs] disabled = true`)
- Set aggressive timeouts (`command_timeout = 500`)
- For git: `[git_status] disabled = true` or increase the scan timeout for large monorepos
- Use `[directory] truncation_length = 3` to avoid expensive path operations

For AI agent/automation contexts where prompts are generated programmatically (in scripts, CI, or agent shells), the overhead matters more than in interactive use. Disabling unnecessary modules gets startup-path overhead under 10ms for most configs.

What shell are you using and what does `starship timings` show as the slow modules?
